core/include/gnuradio-4.0/Graph.hpp

```diff
@@ -780,29 +812,87 @@ struct Graph : public gr::Block<Graph> {
         }
         return allConnected;
     }
+};
 
-    template<std::invocable<BlockModel&> F>
-    void forEachBlockMutable(F&& f) {
-        std::ranges::for_each(_blocks, [f](auto& block_ptr) { std::invoke(f, *block_ptr.get()); });
-    }
+static_assert(BlockLike<Graph>);
 
-    template<std::invocable<Edge&> F>
-    void forEachEdgeMutable(F&& f) {
-        std::ranges::for_each(_edges, f);
-    }
+/*******************************************************************************************************/
+/**************************** Graph helper functions ***************************************************/
+/*******************************************************************************************************/
 
-    template<std::invocable<const BlockModel&> F>
-    void forEachBlock(F&& f) const {
-        std::ranges::for_each(_blocks, [f](auto& block_ptr) { std::invoke(f, std::as_const(*block_ptr.get())); });
-    }
+namespace graph {
 
-    template<std::invocable<const Edge&> F>
-    void forEachEdge(F&& f) const {
-        std::ranges::for_each(_edges, f);
-    }
-};
+namespace detail {
+template<block::Category traverseCategory, typename Fn>
+void traverseSubgraphs(GraphLike auto const& root, Fn&& visitGraph) {
+    using enum block::Category;
 
-static_assert(BlockLike<Graph>);
+    auto recurse = [&visitGraph](GraphLike auto const& graph, auto& self) -> void {
```

Comment:
Accustomed to seeing the recursive `self` as the first arg, but this is fine as well